toolchain/check/modifiers.cpp

```diff
@@ -95,13 +95,18 @@ auto CheckMethodModifiersOnFunction(
     std::optional<SemIR::Inst> parent_scope_inst) -> void {
   if (parent_scope_inst) {
     if (auto class_decl = parent_scope_inst->TryAs<SemIR::ClassDecl>()) {
-      auto inheritance_kind =
-          context.classes().Get(class_decl->class_id).inheritance_kind;
+      auto& class_def = context.classes().Get(class_decl->class_id);
+      auto inheritance_kind = class_def.inheritance_kind;
       if (inheritance_kind == SemIR::Class::Final) {
         ForbidModifiersOnDecl(context, introducer, KeywordModifierSet::Virtual,
                               " in a non-abstract non-base `class` definition",
                               context.insts().GetLocId(parent_scope_inst_id));
       }
+      if (class_def.adapt_id.is_valid()) {
+        ForbidModifiersOnDecl(context, introducer, KeywordModifierSet::Virtual,
+                              " in an adapter definition",
+                              context.insts().GetLocId(parent_scope_inst_id));
+      }
```

Comment:
Do we also reject if the `virtual fn` declaration precedes the `adapt` declaration? (Checking this at the end of the class may allow us to avoid putting the check in two places.)

Reply:
Thanks for the offline help understanding things a bit more.

Implemented a class-block scanning type solution. Though the class's inst block is empty at this point, with some debugging I figured out it was because it hadn't been completed - which happens somewhere where the inst_block_stack is popped and put into the class's inst block.

So I implemented this by iterating over the inst block at the top of the inst block stack. Is that right? Or are there ways the top of the stack could be something other than the class's scope? Or some other reason to do it differently?

(also, currently this means we'll only get one of the various adapter errors - should we change it to get all the errors (fields, virtual functions, base class, etc)?)

Reply:
This looks good, thanks.

Given that the primary error points at the `adapt` declaration, we could consider adding notes for each thing in the class that is incompatible with that (or maybe just each kind of thing -- the first field, first (only) base, and first virtual function). My inclination would be to do that in a different PR if we want to pursue that direction. If so, I think we'd also want to generalize the main error diagnostic text so it works with different kinds of note:
```
error: adapter with state
note: first field is here
note: first virtual function is here
```
or similar might work? (I'm not sure "state" is the right word though.)